cmd/podman/tree.go

```diff
@@ -0,0 +1,175 @@
+	if err != nil {
+		return errors.Wrapf(err, "error while retriving layers of image %q", image.InputName)
+	}
+
+	imageTopLayer := image.TopLayer()
+	ll, ok := layerInfoMap[imageTopLayer]
+	if !ok {
+		return fmt.Errorf("top-layer of image not found in dependency-map")
+	}
+
+	// Create a imageInfo and fill the image and layer info
+	imageInfo := &InfoImage{
+		ID:       image.ID(),
+		RepoTags: image.Names(),
+	}
+	var buffer bytes.Buffer
+
+	if !whatrequires {
+		err := buildImageHierarchyMap(imageInfo, layerInfoMap, ll.ID)
+		if err != nil {
+			return err
+		}
+		err = formatImageHierarchyToBuffer(&buffer, imageInfo)
+		if err != nil {
+			return err
+		}
+	} else {
+		err := formatImageChildrensToBuffer(imageInfo, layerInfoMap, ll.ID, 0, &buffer)
+		if err != nil {
+			return err
+		}
+
+		//Add image info at beginning of buffer.
+		treeData := buffer.String()
+		buffer.Reset()
+		buffer.WriteString(fmt.Sprintf("Image ID     : %s\n", imageInfo.ID[:12]))
+		buffer.WriteString(fmt.Sprintf("RepoTags     : %s\n", imageInfo.RepoTags))
+		buffer.WriteString(fmt.Sprintf("Size         : %v\n", units.HumanSizeWithPrecision(float64(imageInfo.Size), 4)))
+		buffer.WriteString(fmt.Sprintf("Image Layers :\n"))
+		buffer.WriteString(fmt.Sprintf("%s", treeData))
+	}
+
+	fmt.Println(buffer.String())
+
+	return nil
+}
+
+// Stores hierarchy of images such that all parent layers using which image is built are stored in imageInfo
+func buildImageHierarchyMap(imageInfo *InfoImage, layerMap map[string]*image.LayerInfo, layerID string) error {
+	if layerID == "" {
+		return nil
+	}
+	ll, ok := layerMap[layerID]
+	if !ok {
+		return fmt.Errorf("lookup error: layerid  %s, not found", layerID)
+	}
+	if err := buildImageHierarchyMap(imageInfo, layerMap, ll.ParentID); err != nil {
+		return err
+	}
+
+	imageInfo.Layers = append(imageInfo.Layers, *ll)
+	imageInfo.Size = imageInfo.Size + ll.Size
+	return nil
+}
+
+// Stores all children layers which are created using given Image.
+func formatImageChildrensToBuffer(imageInfo *InfoImage, layerMap map[string]*image.LayerInfo, layerID string, prefixSpace int, buffer *bytes.Buffer) error {
+	if layerID == "" {
+		return nil
+	}
+	ll, ok := layerMap[layerID]
+	if !ok {
+		return fmt.Errorf("lookup error: layerid  %s, not found", layerID)
+	}
+	//imageInfo.Layers = append(imageInfo.Layers, *ll)
+	imageInfo.Size = imageInfo.Size + ll.Size
+	for p := 0; p < prefixSpace; p++ {
+		buffer.WriteString(" ")
+	}
+	buffer.WriteString(fmt.Sprintf("└─ ID: %s Size: %v Tags: %s\n", ll.ID[:12], units.HumanSizeWithPrecision(float64(ll.Size), 4), ll.RepoTags))
+	if len(ll.RepoTags) > 0 {
+		prefixSpace++
+	}
+	for _, childID := range ll.ChildID {
+		if err := formatImageChildrensToBuffer(imageInfo, layerMap, childID, prefixSpace, buffer); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func formatImageHierarchyToBuffer(buffer *bytes.Buffer, imageInfo *InfoImage) error {
```

Comment:
It would be nice for the two branches to be consistent about
- printing the image-level ID/tags/size data at the same level of the call stack
- separating, or not, the data collection from the output phase

(… but at this point I’m _personally_ not going to block this PR on this esthetic point. @vrothberg might.)